hazelcast/src/main/java/com/hazelcast/internal/serialization/impl/compact/zeroconfig/ValueReaderWriters.java

```diff
@@ -0,0 +1,1384 @@
+        ARRAY_CONSTRUCTORS.put(Short.class, NullableShortArrayReaderWriter::new);
+        ARRAY_CONSTRUCTORS.put(Short.TYPE, ShortArrayReaderWriter::new);
+
+        CONSTRUCTORS.put(Integer.class, NullableIntegerReaderWriter::new);
+        CONSTRUCTORS.put(Integer.TYPE, IntegerReaderWriter::new);
+        ARRAY_CONSTRUCTORS.put(Integer.class, NullableIntegerArrayReaderWriter::new);
+        ARRAY_CONSTRUCTORS.put(Integer.TYPE, IntArrayReaderWriter::new);
+
+        CONSTRUCTORS.put(Long.class, NullableLongReaderWriter::new);
+        CONSTRUCTORS.put(Long.TYPE, LongReaderWriter::new);
+        ARRAY_CONSTRUCTORS.put(Long.class, NullableLongArrayReaderWriter::new);
+        ARRAY_CONSTRUCTORS.put(Long.TYPE, LongArrayReaderWriter::new);
+
+        CONSTRUCTORS.put(Float.class, NullableFloatReaderWriter::new);
+        CONSTRUCTORS.put(Float.TYPE, FloatReaderWriter::new);
+        ARRAY_CONSTRUCTORS.put(Float.class, NullableFloatArrayReaderWriter::new);
+        ARRAY_CONSTRUCTORS.put(Float.TYPE, FloatArrayReaderWriter::new);
+
+        CONSTRUCTORS.put(Double.class, NullableDoubleReaderWriter::new);
+        CONSTRUCTORS.put(Double.TYPE, DoubleReaderWriter::new);
+        ARRAY_CONSTRUCTORS.put(Double.class, NullableDoubleArrayReaderWriter::new);
+        ARRAY_CONSTRUCTORS.put(Double.TYPE, DoubleArrayReaderWriter::new);
+    }
+
+    private ValueReaderWriters() {
+    }
+
+    /**
+     * Returns the reader writer for the given {@code type}.
+     *
+     * @param clazz       Top level class
+     * @param type        Class to return the reader writer for
+     * @param genericType Generic type of the {@code type}
+     * @param fieldName   Name of the field
+     * @return Appropriate reader for the given {@code type}
+     */
+    public static ValueReaderWriter<?> readerWriterFor(Class<?> clazz, Class<?> type, Type genericType, String fieldName) {
+        if (type.isArray()) {
+            Class<?> componentType = type.getComponentType();
+            return createReaderWriterForArray(clazz, componentType, fieldName);
+        } else if (isList(type)) {
+            Class<?> componentType = getSingleComponentType(genericType);
+            ValueReaderWriter readerWriter = createReaderWriterForArray(clazz, componentType, fieldName);
+            return new ArrayListReaderWriter(fieldName, componentType, readerWriter);
+        } else if (isSet(type)) {
+            Class<?> componentType = getSingleComponentType(genericType);
+            ValueReaderWriter readerWriter = createReaderWriterForArray(clazz, componentType, fieldName);
+            return new HashSetReaderWriter(fieldName, componentType, readerWriter);
+        } else if (isMap(type)) {
+            BiTuple<Class<?>, Class<?>> componentTypes = getTupleComponentTypes(genericType);
+            ValueReaderWriter keyReaderWriter
+                    = createReaderWriterForArray(clazz, componentTypes.element1, fieldName + "Keys");
```

Comment:
Asking to be sure, if there are fields `A` and `AKeys` in a class where A is a HashMap they won't collide right?

Reply:
Yeah, good catch. I will change it to `!keys` and `!values` to avoid possibility of collision as `!` can't be used in identifier names